File: MergeSort.java
import java.util.Scanner;
//This program contains an implementation of Merge sort in the main func where a user defined array is sorted using Merge Sort.
public class MergeSort {

    void merge(char arr[], int l, int m, int r) {
        int n1 = m - l + 1;
        int n2 = r - m;
        char L[] = new char[n1];
        char R[] = new char[n2];
        for (int i = 0; i < n1; ++i){
            L[i] = arr[l + i];
	  }
        for (int j = 0; j < n2; ++j){
            R[j] = arr[m + 1 + j];
	  }
        int i = 0, j = 0;     
        int k = l;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }
        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }
        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }
    void sort(char arr[], int l, int r) {
        if (l < r) {
            int m = l + (r - l) / 2;
            sort(arr, l, m);
            sort(arr, m + 1, r);            
            merge(arr, l, m, r);
        }
    }

    static void printArray(char arr[]) {
        int n = arr.length;
        for (int i = 0; i < n; ++i)
            System.out.print(arr[i] + " ");
        System.out.println();
    }

    public static void main(String args[]) {
        Scanner in = new Scanner(System.in);
        int n;
        System.out.println("Enter n");
        n = in.nextInt();

        char arr[] = new char[n];

        System.out.println("Enter Array");
        for (int i = 0; i < n; i++) {
            arr[i] = in.next().charAt(0);
        }

        System.out.println("Given Array");
        printArray(arr);

        MergeSort ob = new MergeSort();
        ob.sort(arr, 0, arr.length - 1);

        System.out.println("\nSorted array");
        printArray(arr);
    }
}

File: DSA in Python/merge_sort.py
def merge(arr, l, m, r):
    n1 = m - l + 1
    n2 = r - m
    L = arr[l:m + 1]
    R = arr[m + 1:r + 1]
    
    i = j = 0
    k = l
    
    while i < len(L) and j < len(R):
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1
    
    while i < len(L):
        arr[k] = L[i]
        i += 1
        k += 1
    
    while j < len(R):
        arr[k] = R[j]
        j += 1
        k += 1

def merge_sort(arr, l, r):
    if l < r:
        m = l + (r - l) // 2
        merge_sort(arr, l, m)
        merge_sort(arr, m + 1, r)
        merge(arr, l, m, r)

def print_array(arr):
    print(" ".join(arr))

if __name__ == "__main__":
    n = int(input("Enter n: "))
    arr = []

    print("Enter array elements:")
    for _ in range(n):
        arr.append(input().strip())

    print("Given Array:")
    print_array(arr)

    merge_sort(arr, 0, len(arr) - 1)

    print("\nSorted Array:")
    print_array(arr)
